server/src/main/java/org/elasticsearch/indices/SystemIndexMappingUpdateService.java

```diff
@@ -234,7 +234,7 @@ static State calculateIndexState(ClusterState state, SystemIndexDescriptor descr
             return null;
         }
 
-        final boolean isIndexUpToDate = INDEX_FORMAT_SETTING.get(indexMetadata.getSettings()) == descriptor.getIndexFormat();
+        final boolean isIndexUpToDate = INDEX_FORMAT_SETTING.get(indexMetadata.getSettings()) == descriptor.getIndexFormat().version();
```

Comment:
I would like to clarify what the intention is when this PR is complete.

At present the `State` distinguishes between an upgrade being needed and a mappings update being required.

An upgrade is indicated by a change in `indexFormat` and is considered to mean that complex plugin-specific code needs to run outside of the system index upgrader: https://github.com/elastic/elasticsearch/blob/f1bc2e0f2f51c0e9cf8c254f796bc75ecb17b7d9/server/src/main/java/org/elasticsearch/indices/SystemIndexMappingUpdateService.java#L176-L182

A mappings update is considered much simpler, and is currently indicated by the version field in the metadata incrementing.

This PR seems to be merging both concepts into `IndexFormat`.

I notice there isn't a place in the code where `UpgradeStatus.NEEDS_UPGRADE` is actually used. So is the intention to scrap that functionality and instead use `IndexFormat` to indicate the need for a mappings update?

Reply:
First of all, your comment is very helpful and clarifying. Thank you for reviewing this!

Here's how I've been thinking about how we need to manage system indices:

| Issue | Old way of doing things | Where this PR is heading |
|--|--|--|
| Choosing mappings in mixed cluster | Choose prior descriptor with mapping version matching minimum node | Determine minimum index format for this descriptor across all nodes, and choose the prior descriptor with index format matching that |
| When to update mappings? | If we have a non-mixed-version cluster and the index format matches the descriptor | Determine minimum index format for this descriptor across all nodes, and update mappings if it equals the master node's descriptor format |
| When to invoke POST /_migration/system_features | Index version before global "no migration required" constant | The descriptor's minimum `IndexVersion` is later than the actual `IndexVersion` of the system index |

I see that I've assumed that a breaking change in mappings can only occur when there's a change in the `IndexFormat`, and that all such changes are best handled through the system feature migration API (which had a main use case of avoiding Lucene incompatibility). 

In order to account for breaking changes in mappings where there hasn't been a change in `IndexVersion` (which could happen due to feature development), we might need the following concepts:
* Descriptor mapping version: incremented for additive changes, used in prior descriptor system, replaces use of Version with an integer
* Descriptor index format version: incremented for breaking changes (rename to avoid confusion with minimum `IndexVersion`?)
* Descriptor minimum `IndexVersion`: used with system feature migration system

This would give us:

| Issue | Another way of doing things |
|--|--|
| Choosing mappings in mixed cluster | Determine minimum descriptor mapping version across all nodes, and choose the prior descriptor with mapping version matching that |
| When to update mappings? | When the index's mapping version is before the cluster's minimum descriptor mapping version for that index (but descriptor index format versions match) |
| When to invoke POST /_migration/system_features | The descriptor's minimum `IndexVersion` is later than the actual `IndexVersion` of the system index |

Have I missed anything important here? Would the mapping version / format version / `IndexVersion` system address your concerns?

Reply:
For ML we currently update mappings more aggressively than the system index manager. We update them early in any action that might later index a document containing the new fields. So ML index mappings might get updated to a new version when only one node in the cluster is on that new version. But that works because the newer mappings only ever add fields, never remove them, and extra mapped fields that aren't populated are not a problem.

An example is: https://github.com/elastic/elasticsearch/blob/69914bf61e40994d561f168b81bc2375208c138e/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/JobManager.java#L279

Further down this calls: https://github.com/elastic/elasticsearch/blob/6592b79560f57270d5d51f55d215a3902e5b280f/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/persistence/ElasticsearchMappings.java#L103

That approach is used for both ML system indices and ML hidden indices.

Currently that code is using the same version meta field that the system index manager uses. After this PR is merged it would make sense for the ML code to be updated to use the new mechanism of the system index manager, for ML hidden indices as well as ML system indices.

So for the "When to update mappings?" question, ML doesn't mind whether you go with the "Where this PR is heading" approach or the "Another way of doing things" approach. You can do whatever works best for all the non-ML system indices.

The key requirement for ML is that we can continue to update our index mappings separately to the system index manager. It's OK if the need for this is detected by an increment to the `index.format` setting on the index. One side effect of that is that index mappings updates will also need to put new index settings whereas today putting new mappings with a `_meta` field is enough to both update the mappings and indicate that they are up to date. But if that's the way this work is going it's not a big problem. I'd rather this work got finished quickly so ML can build on top of it soon rather than cause a delay to this work by changing approach part way through.

The other requirement for ML is that it must always be tolerable to have a `_meta.version` field in the mappings, even if the latest system index manager and latest ML code don't use it. This is because in a mixed version cluster that includes version 8.9 or older, if an 8.9 or older node sees no `_meta.version` field on the index mappings then it will think they're _really_ old and update them. We don't want old and new nodes to fight like this, and we could avoid it by leaving `_meta.version` as 8.9.99 (or some other suitable version) in the mappings forever.